### PR TITLE
#145 Implement Minecraft player list component

### PR DESCRIPTION
**Pull request by kevinzhangTT**
# Description

https://github.com/Lodestone-Team/lodestone/issues/145

Implements a Minecraft player list component that displays the player's username and avatar picture, and updates in real time as players join or leave the server. The profile pictures are retrieved from https://mc-heads.net/#basic using the player's UUID. The NAME button above the list allows for sorting usernames in alphabetical or reverse alphabetical order.

image

If there is no one online, the text changes to say "No players online".

image

The main files that contribute to this PR are are [MinecraftPlayerList.tsx](https://github.com/Lodestone-Team/lodestone/compare/0.4.3...145-0.4.3-player-list#diff-22ccce3d6d44067789bc988096cf61077cbe6acd43fb466403207181d741242b) and [PlayerListCard.tsx](https://github.com/Lodestone-Team/lodestone/compare/0.4.3...145-0.4.3-player-list#diff-8866a9890a7b4622468f92c6cc21296f81121ee96dccd918d7308c5946c6e5f6). MinecraftPlayerList.tsx mainly handles player data like setting the avatar and tracking online players, while PlayerListCard.tsx contains the UI.

[InstanceInfo.ts](https://github.com/Lodestone-Team/lodestone/compare/0.4.3...145-0.4.3-player-list#diff-a1480f44ddfd73db4591896dfd5eacb897e8f2575999d4acc06c04974ab3c0e4) and [EventStream.ts](https://github.com/Lodestone-Team/lodestone/compare/0.4.3...145-0.4.3-player-list#diff-ce81eee6c4968be3bd41724436bc659c835735ce4c9fd4a20ac190885289f084) also had to be modified in some way to support `Player` objects and updating the list of them to reflect in the UI in real time.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Extremely thorough testing B)

minecraft-player-list-test

### CODE REVIEW

File: src/components/Minecraft/MinecraftPlayerList.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import { faArrowDown, faArrowUp } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+// import { Player } from 'bindings/Player';
+
+export default function MinecraftPlayerList() {
+
+  const { selectedInstance: instance } = useContext(InstanceContext);
+  const uuid = instance?.uuid;
+  
+  // JUST FOR TESTING
```
Comment (joeywangzr) [inline src/components/Minecraft/MinecraftPlayerList.tsx:14]: remove these comments

**Review by joeywangzr — COMMENTED** [review]
otherwise lgtm

File: src/components/Minecraft/MinecraftPlayerList.tsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  //   { type: 'MinecraftPlayer', uuid: 'c709ad04cf80493581db9bf715c92733', name: '3nj' },
+  //   { type: 'MinecraftPlayer', uuid: 'c1d2e3157c3a4a978a8effd188cad54d', name: 'PocketSage' },
+  //   { type: 'MinecraftPlayer', uuid: 'b65635533dff4ed0a4e6c24032fc146e', name: 'Kevin1' },
+  //   { type: 'MinecraftPlayer', uuid: 'e0688b364e92478fb8b7fcabcb811f7e', name: 'kevin2' },
+  //   { type: 'MinecraftPlayer', uuid: '66aa923f94ae42fba61fbcccd642a4c1', name: 'SpookyCrepe' }
+  // ];
+
+  // Default value of playerList will be an empty array since we cannot have useState after a conditional block
+  const [playerList, setPlayerList] = useState([
+    { type: 'MinecraftPlayer', uuid: '', name: ''}
```
Comment (Ynng) [inline src/components/Minecraft/MinecraftPlayerList.tsx:30]: this should be an empty array

File: src/components/Minecraft/MinecraftPlayerList.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  ]);
+  
+  // Ascending order is referring to alphabetical order
+  const [sortOrder, setSortOrder] = useState<'asc' | 'desc'>('asc');
+
+  // Update playerList every time there a player leaves or joins
+  useEffect(() => {
+    console.log('changed detected from useEffect');
+    if (instance && instance.player_list) {
+      setPlayerList(instance.player_list);
```
Comment (Ynng) [inline src/components/Minecraft/MinecraftPlayerList.tsx:40]: why do we have this state in the middle?
if it's for sorting, I believe we can use a `useMemo` instead

File: src/components/Minecraft/MinecraftPlayerList.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            </h1>
+          </div>
+        </div>
+      </div>
+    );
+  }
+  
+  // API to get the avatar head png 16x16 px
+  const mcHeadURL = 'https://mc-heads.net/avatar/';
+  const avatarDimension = 128;
```
Comment (Ynng) [inline src/components/Minecraft/MinecraftPlayerList.tsx:77]: do we really need 128?

**Review by Ynng — CHANGES_REQUESTED** [review]
overall really nice, just a few small things
mostly regarding the player list use state

**Comment by Ynng** [comment]
also the icon looks a bit small... can you @Arcslogger take a look next week?

**Review by Ynng — APPROVED** [review]
code looks good! let's merge this pr.


@CheatCod when can we get the player_list api changed merged to the backend?
https://github.com/Lodestone-Team/lodestone_core/tree/player_list_in_instance_info

**Comment by CheatCod** [comment]
backend will follow up once the configurable refactor is merged in

**Pull request merged**